### PR TITLE
Arrow string array dtype

### LINKED ISSUE #35169: Plan for a native string dtype
Apache Arrow has support for natively storing UTF-8 data. And [work is ongoing](https://github.com/apache/arrow/pull/7449)
adding kernels (e.g. `str.isupper()`) to operate on that data. This issue
is to discuss how we can expose the native string dtype to pandas' users.

There are several things to discuss:

1. How do users opt into this behavior?
2. A fallback mode for not implemented kernels.

### How do users opt into Arrow-backed StringArray?

The primary difficulty is the additional Arrow dependency. I'm assuming that we
are not ready to adopt it as a required dependency, so all of this will be
opt-in for now (though this point is open for discussion).

StringArray is marked as experimental, so our usual API-breaking restrictions
rules don't apply. But we want to do this in a way that's not too disruptive.

There are three was to get a `StringDtype`-dtype array today:

1. Infer: `pd.array(['a', 'b', None])`
2. Explicit `dtype=pd.StringDtype()`
3. String alias `dtype="string"`

My preference is for all of these to stay consistent. They all either give a
StringArray backed by an object-dtype ndarray or a StringArray backed by Arrow
memory.

I also have a preference for not keeping around our old implementation for too
long. So I don't think we want something like `pd.PythonStringDtype()` as a
way to get the StringArray backed by an object-dtype ndarray.

The easiest way to support this is, I think, an option.

```python
>>> pd.options.mode.use_arrow_string_dtype = True
```

Then all of those would create an Arrow-backed StringArray.

### Fallback Mode

It's likely that Arrow 1.0 will not implement all the string kernels we need.
So when someone does

```python
>>> Series(['a', 'b'], dtype="string").str.normalize()  # no arrow kernel
```

we have a few options:

1. Raise, stating that there's no kernel for normalize.
2. PerformanceWarning, astype to object, do the operation, and convert back

I'm not sure which is best. My preference for now is probably to raise, but I could see doing either.

### PR DESCRIPTION
**Pull request by TomAugspurger**
This is a continuation of https://github.com/pandas-dev/pandas/pull/35259, updated to use a single `StringDtype` parameterized by `storage` (python or pyarrow). We do use different classes for the array though.

I wouldn't recommend reviewing this yet. In particular, I need to refactor the string ops implementation. That will likely be done in a separate PR to keep things smaller here.

```python
In [1]: import pandas as pd

In [2]: a = pd.Series(pd.array([pd.NA, pd.NA], dtype=pd.StringDtype("python")))

In [3]: b = pd.Series(pd.array([pd.NA, pd.NA], dtype=pd.StringDtype("pyarrow")))

In [4]: a.array
Out[4]:
<StringArray>
[<NA>, <NA>]
Length: 2, dtype: StringDtype[python]

In [5]: b.array
Out[5]:
<ArrowStringArray>
[<NA>, <NA>]
Length: 2, dtype: StringDtype[pyarrow]
```

Closes https://github.com/pandas-dev/pandas/issues/35169

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Just FYI @xhochy.

**Comment by xhochy** [comment]
@TomAugspurger I would continue to work on this starting next week again. Do we want to coordinate somehow?

File: pandas/core/arrays/base.py [hunk 1]
```diff
@@ -459,6 +461,7 @@ def astype(self, dtype, copy=True):
         from pandas.core.arrays.string_ import StringDtype
 
         dtype = pandas_dtype(dtype)
+        # FIXME: Really hard-code here?
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/base.py:464]: This needs the general 'astype' machinery refactor to fix this, I think

File: pandas/core/arrays/string_.py [hunk 2]
```diff
@@ -50,28 +52,99 @@ class StringDtype(ExtensionDtype):
...
+        string : str
+            The type of the name. The storage type will be taking from `string`.
+            Valid options and their storage types are
+
+            ========================== ==============
+            string                     result storage
+            ========================== ==============
+            ``'string'``               global default
+            ``'string[python]'``       python
+            ``'StringDtype[python]'``  python
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/string_.py:92]: Why allow the full class name "StringDtype" as string as well? (I don't think we do that for other dtypes?)

EDIT: ah, I see you used that as name above. But so then: why not "string" as name?

File: pandas/core/arrays/string_.py [hunk 3]
```diff
@@ -50,28 +52,99 @@ class StringDtype(ExtensionDtype):
...
+            storage = get_option("mode.string_storage")
+        if storage not in {"python", "pyarrow"}:
+            raise ValueError(
+                f"Storage must be 'python' or 'pyarrow'. Got {storage} instead."
+            )
+        self.storage = storage
+
+    @property
+    def name(self):
+        return f"StringDtype[{self.storage}]"
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/string_.py:70]:
```suggestion
        return f"string[{self.storage}]"
```

?

File: pandas/core/arrays/string_arrow.py [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+        The string methods are available on Series backed by
+        a ArrowStringArray.
+
+    Notes
+    -----
+    ArrowStringArray returns a BooleanArray for comparison methods.
+
+    Examples
+    --------
+    >>> pd.array(['This is', 'some text', None, 'data.'], dtype="arrow_string")
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/string_arrow.py:73]:
```suggestion
    >>> pd.array(['This is', 'some text', None, 'data.'], dtype="string[arrow]")
```

File: pandas/core/arrays/string_arrow.py [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+            self.data = values
+        else:
+            raise ValueError(f"Unsupported type '{type(values)}' for ArrowStringArray")
+        self._dtype = StringDtype(storage="pyarrow")
+
+    @classmethod
+    def _from_sequence(cls, scalars, dtype=None, copy=False):
+        # TODO(ARROW-9407): Accept pd.NA in Arrow
+        scalars_corrected = [None if isna(x) else x for x in scalars]
+        return cls(pa.array(scalars_corrected, type=pa.string()))
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/string_arrow.py:92]:
```suggestion
        return cls(pa.array(scalars_corrected, type=pa.string(), from_pandas=True))
```

then np.nan and pd.NA will be accepted as "null" as well, in addition to None

**Review by jorisvandenbossche — COMMENTED** [review]
(I know you said to not yet review, but was just curious ;) The dtype dispatching to two different array classes looks good to me)

File: pandas/core/arrays/string_.py (hunk 3 above)
Comment (TomAugspurger) [reply pandas/core/arrays/string_.py:70]: Agreed, this seems better.

**Comment by TomAugspurger** [comment]
Closing in favor of https://github.com/pandas-dev/pandas/pull/35259 (hopefully)

**Comment by simonjayhawkins** [comment]
@TomAugspurger @xhochy OK to merge these changes into #35259?

**Comment by TomAugspurger** [comment]
I'm not sure if we have complete agreement over how this will be exposed to the user yet in https://github.com/pandas-dev/pandas/issues/35169. This PR was taking a strong opinion on

1. One user-facing `pd.StringDtype[storage]`, parametrized by storage = `python/pyarrow`.
2. One or two (or three) user-facing  StringArray classes, an `ArrowStringArray`, `PythonStringArray`, and maybe a base `StringArray`.

Given the uncertainty around the API here, I'd prefer to just move forward with an internals-only `ArrowStringArray` and decide on the exact API later (unless we can agree on something quickly).

**Comment by simonjayhawkins** [comment]
> I'd prefer to just move forward with an internals-only `ArrowStringArray`

OK, i've just changed #35259 so that the doctest passed. I should probably revert that change and change the doctest so that ArrowStringDtype is not a top level import.

**Comment by jorisvandenbossche** [comment]
Personally, I think there is enough agreement to at least start making a concrete PR, which will more easily trigger further discussion on it. 
(but let's bring this up again in the issue).

**Comment by jorisvandenbossche** [comment]
That said, I think it's certainly fine to focus on the actual implementation in #35259, and leave the user exposing API for a separate PR. There is already more than enough to review and discuss for the internal implementation details ;)

**Comment by simonjayhawkins** [comment]
I don't seem to able to push to this branch. will open a new PR

**Comment by simonjayhawkins** [comment]
> I don't seem to able to push to this branch. will open a new PR

#39908

**Pull request closed without merging**